design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+use the new APIs without any other system. General sentiment was that it is important for users to
+understand the concept of `PARTITIONS` soon into their journey understanding Kafka. There are four
+alternatives: 
+  * Do not require `PARTITIONS`/`REPLICAS` to create the kafka topic. This is a slight change that
+  will make it easier to experiment, but may open the door to accidentally creating kafka topics.
+  With the release of 5.2, this will also not be consistent with CSAS/CTAS, which default to the
+  source replication/partitioning instead of using some default configured values.
+  * Introduce another `CREATE TOPIC` API which will create empty topics. I believe this will cause
+  confusion, as it departs farther from SQL standard and puts `TOPIC` on the same level as the
+  duality of Tables/Streams (which it is not, it is a physical manifestation of a stream/table).
+  * Don't fail when creating streams/topics with non-existing topics, and create those topics if
+  a call to `INSERT INTO ...` is called with a stream/table target that does not exist. This seems
+  more "magical" and runs into issues when specifying partition/replica counts.
+  * Introduce another `REGISTER TABLE/STREAM` to represent the difference. This moves away from
+  SQL standard and adds more complexity for the user to understand.
+
+### Code Enhancements
+
+KSQL in interactive mode is already configured to produce data to a command topic. This KLIP can
+leverage that existing functionality, which will require piping through various configurations into
+the `ServiceContext`. Beyond that, we must introduce a new command and Statement type to reflect the
+new functionality. 
+
+In interactive mode, this will be handled without distributing the statement to other servers. This
+operation will not be available in headless or embedded mode for the first version of the feature.
+
+### Schema Validation
+
+`INSERT INTO ... VALUES` will only support inserting into values for existing KSQL Streams/Tables,
+which already have a schema associated with them. Before producing a record to the underlying topic,
+the values for the fields will be verified against the schema. 
+
+### Serialization Support
+
+JSON, AVRO and DELIMITED value formats will all be supported, the latter of which will not accept
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+If the user specifies `AVRO` encoding in their `CREATE` statement, we will register the topic with 
+the schema registry proactively if the topic did not previously exist. Note that this will not 
+change how we handle `CREATE` statements for topics that already exist (today we do not validate
+that the schema supplied matches the schema in Schema Registry).
```

Comment:
> (today we do not validate that the schema supplied matches the schema in Schema Registry).

Ermm... are you sure? I know there's code in KSQL that will fail if the schema is not evolvable. It's possible that is only the case for CSAS/CTAS statements at present, but even if that is the case we should extend it to cover CS/CT statements. Why? Well, up until the point we introduce INSERT functionality there was no way to insert a value into a stream registered via a CT/CS statement, so the lack of schema validation could never result in KSQL producing records to the topic with an incompatible schema.  With the addition of `INSERT` this is now possible, so it is important that we validate schema compatibility.

Reply:
> Ermm... are you sure?

I tested it locally issuing both of these:

Reply:
The test I ran was to run the following statements:

Reply:
Yeah, as I said, we certainly should be validating now that we have `INSERT`.